woocommerce/woocommerce
# Re-introduce 44084 and 44790 with back-compat

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





We reverted #44080  and #44790 in #44876 earlier after we discovered some backward compatibility issues, given that after removing the REST API, we were not calling the `rest_request_after_callbacks` anymore. This PR aims to re-introduce those changes with appropriate backward compatibility.

Note that we call here `rest_request_after_callbacks` manually, which is generally called in a REST request lifecycle. While we do it here for backward compatibility, it is also important to note that we are not calling other REST lifecycle hooks such as `rest_request_before_callbacks` or `rest_dispatch_request`. Upon auditing usage of these hooks, it looks like only `rest_request_after_callbacks` is used in the context of changing the response of an API request, which is relevant to us.

Alternatively, we can choose to not add this hook, in which case the 3pd plugins currently hoping to change the store API data would need to find some other way to modify this information.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Similar to previous PRs, this needs exploratory testing, especially on cart and checkout blocks. To test, complete a checkout using cart and checkout blocks on any theme.

Additionally, we should also test the back compat and verify if the `rest_request_after_callbacks` filter is being executed like so:

1. Add this as a code snippet to hook on to the rest filter:
```php
<?php
/**
 * Plugin Name: API Profiler
 */

add_filter( 'rest_request_after_callbacks', function () {
	$e = new \Exception;
	js_console_log( array(
		'message' => 'REST API after callback',
		'stack' => $e->getTraceAsString(),
	));

} );

function js_console_log( $data ) {
	?>
	<script>
		console.log( <?php echo json_encode( $data ); ?> );
	</script>
	<?php
}
```



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
